File: src/bin/0735-asteroid-collision.py
# @leetup=info id=735 lang=python3 slug=asteroid-collision



from typing import Optional

# @leetup=code
class Solution:
    def asteroidCollision(self, asteroids: list[int]) -> list[int]:
        res = []
        for s in asteroids:
            while res and s < 0 < res[-1]:
                if res[-1] == -s:
                    res.pop()
                    s = 0
                elif res[-1] > -s:
                    s = 0
                else:
                    res.pop()
            if s:
                res.append(s)
        return res


# @leetup=code

File: src/bin/0735-asteroid-collision.rs
// @leetup=info id=735 lang=rust slug=asteroid-collision

// @leetup=code
impl Solution {
    pub fn asteroid_collision(asteroids: Vec<i32>) -> Vec<i32> {
        let mut res = Vec::new();
        for mut s in asteroids {
            while let Some(&last) = res.last() {
                if last < 0 || s >= 0 {
                    break;
                }
                if last == -s {
                    res.pop();
                    s = 0;
                    break;
                } else if last > -s {
                    s = 0;
                    break;
                } else {
                    res.pop();
                }
            }
            if s != 0 {
                res.push(s);
            }
        }
        res
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::asteroid_collision;
}
